### DISCUSSION TITLE
Connections created at runtime hang when used

### DISCUSSION THREAD

**Discussion by pilord12** [apache/airflow#34149] [Q&A]
I am trying to dynamically set up a connection to AWS based on params provided to my DAG.  The connection seemingly gets created and added appropriately; however, the "create_buckets" task hangs when I try to use the connection.  The same hang occurs if I try to do `conn.test_connection()` from within the `setup_aws_connection` task.

If it's relevant, [this StackOverflow response](https://stackoverflow.com/questions/53740885/create-and-use-connections-in-airflow-operator-at-runtime) was the basis for my code.

Any ideas what I could be missing?

```
import pendulum

from airflow import DAG
from airflow.decorators import task
from airflow.providers.amazon.aws.operators.s3 import S3CreateBucketOperator, S3DeleteBucketOperator
from airflow.models.param import Param
from airflow.models.connection import Connection
from airflow.operators.python import get_current_context
from airflow import settings

with DAG(
        dag_id="provision-assets",
        schedule=None,
        start_date=pendulum.datetime(2023, 8, 1),
        params={
            "aws_connection_id": Param(
                type="string",
                description="Name of AWS connection to be stored in Airflow's state",
                default="localstack_dev",
            ),
            "aws_endpoint_url": Param(
                type="string",
                description="URL of AWS instance to hit",
                default="http://localhost/localstack",
            ),
            "aws_region": Param(
                type="string",
                description="AWS region to use",
                default="us-east-1",
            )
        },
        render_template_as_native_obj=True
) as my_dag:
    @task
    def setup_aws_connection():
        context = get_current_context()
        conn = Connection(
            conn_id=context["params"]["aws_connection_id"],
            conn_type="aws",
            extra={
                # Specify extra parameters here
                "region_name": context["params"]["aws_region"],
                "endpoint_url": context["params"]["aws_endpoint_url"],
                "test_endpoint_url": context["params"]["aws_endpoint_url"],
            },
        )

        session = settings.Session()

        existing_conn = session.get(Connection, conn.conn_id)

        if existing_conn is not None:
            session.add(conn)
            session.commit()

    a = S3CreateBucketOperator(task_id="create_buckets", aws_conn_id="localstack_dev",
                                       dag=my_dag, bucket_name="foo")
    aws_conn >> a
```

**Comment by potiuk** [comment]
IT's a bad idea to generate connections dynamically like that - this will for example stop working with AIP-44 (internal DB API) . 

Creating Connection object like that is not something that is "public interface" of Airlfow and likely this is the reason why it does not work (and it's wrong to assume that even if you fix it, it will remain as is). 

I'd advise you to use API to create connections https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/post_connection

**Reply by pilord12** [reply]
Okay, gotcha, fair enough.  I can use the API to create a connection in my application code before triggering Airflow.  With that, I can simplify my DAG to:

```
with DAG(
        dag_id="provision-assets",
        schedule=None,
        start_date=pendulum.datetime(2023, 8, 1),
        params={
            "aws_connection_id": Param(
                type="string",
                description="Name of AWS connection to be stored in Airflow's state",
                default="localstack_dev",
            ),
        },
        render_template_as_native_obj=True
) as my_dag:
    @task
    def conn_id():
        context = get_current_context()
        return context["params"]["aws_connection_id"]

    cid = conn_id()

    a = S3CreateBucketOperator(task_id="create_buckets", aws_conn_id=cid,
                                       dag=my_dag, bucket_name="foo")

    cid >> a
```

however, I still see the same behavior--hanging upon trying to use the AWS connection.  Not much in the logs:

```
[2023-09-07, 14:33:10 UTC] {task_command.py:410} INFO - Running <TaskInstance: provision-assets.create_buckets manual__2023-09-07T14:32:51+00:00 [running]> on host Keatons-MacBook-Pro.local
[2023-09-07, 14:33:10 UTC] {taskinstance.py:1545} INFO - Exporting env vars: AIRFLOW_CTX_DAG_OWNER='airflow' AIRFLOW_CTX_DAG_ID='provision-assets' AIRFLOW_CTX_TASK_ID='create_buckets' AIRFLOW_CTX_EXECUTION_DATE='2023-09-07T14:32:51+00:00' AIRFLOW_CTX_TRY_NUMBER='1' AIRFLOW_CTX_DAG_RUN_ID='manual__2023-09-07T14:32:51+00:00'
[2023-09-07, 14:33:10 UTC] {base.py:73} INFO - Using connection ID 'localstack_dev' for task execution.
```

Am I missing something wrt using the connection?  Previously it worked when I exported connection details as environment variables, but IIUC I should be able to connect without using env variables (which I would like to do, since I would like to specify the connection ID at runtime).

**Reply by potiuk** [reply]
no idea. But It's likely something environmental - maybe it asks you for user/password when you try to connect (in non-interactive mode?) wild guess. My guess is that your connection is not enough to make the connection works - maybe you have some typos?

**Reply by pilord12** [reply]
Gotcha, so conceptually you'd expect the above code to work with a valid connection?  Let me dig into that piece.

**Reply by pilord12** [reply]
The odd thing is that if I do `airflow connections test localstack_dev`, the testing succeeds.  I would expect that to mean that the connection is valid, or is there more to it?

**Reply by Taragolis** [reply]
Please check important section in top of [Amazon Web Services Connection](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/aws.html) documentation. I guess you might found answer for you question.

**Reply by pilord12** [reply]
Hmm, okay, so the connection test is not necessarily sufficient to determine if the connection will work, fair enough.  To narrow down the issue, I decided to go back to generating the connection statically for the time being.  The following snippet works:

```
conn = Connection(
    conn_id="localstack_dev",
    conn_type="aws",
    extra={
        # Specify extra parameters here
        "region_name": "us-east-1",
        "endpoint_url": "http://localhost/localstack",
        "test_endpoint_url": "http://localhost/localstack",
        "aws_access_key_id": "test",
        "aws_secret_access_key": "test",
    },
)
# Generate Environment Variable Name and Connection URI
env_key = f"AIRFLOW_CONN_{conn.conn_id.upper()}"
conn_uri = conn.get_uri()
print(f"{env_key}={conn_uri}") # prints the same value as is retrieved with airflow connections list

os.environ[env_key] = conn_uri
print(conn.test_connection())  # Validate connection credentials.
```

but if I no longer create the connection in code and instead only set the environment variables to the same values (see snippet below), the connection hangs.  The reason I perform this test is to try to set environment variables to what I see when I run `airflow connections list`, which IIUC should be what is used when I specify `conn_id` to the operator.  It seems like there's something I'm missing which contributes to connection formation, or maybe `airflow connections list` doesn't show all the information?

```
# Generate Environment Variable Name and Connection URI
env_key = f"AIRFLOW_CONN_LOCALSTACK_DEV"
conn_uri = "aws:///?region_name=us-east-1&endpoint_url=http%3A%2F%2Flocalhost%2Flocalstack&test_endpoint_url=http%3A%2F%2Flocalhost%2Flocalstack&aws_access_key_id=test&aws_secret_access_key=test"
print(f"{env_key}={conn_uri}")

os.environ[env_key] = conn_uri
```

**Reply by Taragolis** [reply]
AFAIK `airflow connections list` only works with connections which are stored into the DB backend.
In theory it could be changed, by adding additional methods to SecretsBackends however it required quite a few rework in different places.

In addition I want to add, that environment variables should be provided during Airflow workers/scheduler start. If you create it in one task it won't be visible in another task. Each task is created in a separate process and only has access to a copy of env vars of the parent process.

**Reply by pilord12** [reply]
Right, but isn't the DB backend where I am looking for connections when I specify `aws_conn_id`?

Yes, understood--I am setting env variables outside all tasks.

**Reply by potiuk** [reply]
No. The sequence of looking for connection is:

a) env variable
b) secret backends
c) metadata store

See documentation that explains all of that comprehensively https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html

**Reply by potiuk** [reply]
Only metadata connections are shown with `airflow connections list`

**Reply by pilord12** [reply]
Okay, I was aware of the three sources but wasn't aware of the defined order of their use.  My confusion still remains, though--how do I use a connection from within a DAG which was not created in Python code in that file?

**Reply by potiuk** [reply]
> Okay, I was aware of the three sources but wasn't aware of the defined order of their use. My confusion still remains, though--how do I use a connection from within a DAG which was not created in Python code in that file?

I am confused what you are asking for.

**Reply by pilord12** [reply]
Sorry, let me try to clarify.  If I create a connection in Python and export the connection string as an environment variable (see my code snippet in [this reply](https://github.com/apache/airflow/discussions/34149#discussioncomment-6942163)), things seem to work as expected.

However, if I create seemingly the same connection by POST to the Airflow API or by adding it via `airflow connections add`, the connection hangs when I attempt to use it in the DAG.  I'm confused why this is the case and feel like I'm missing something fundamental about how to reference connections that are created externally to the DAG file.

**Reply by potiuk** [reply]
I am equally confused. No idea why it hangs. Most likely it's some typo on your side - but you should debug it.
